openlr/candidate_paths_getter.cpp

```diff
@@ -0,0 +1,339 @@
+{
+  ScopedTimer t(m_stats.m_startLinesTime);
+
+  for (auto const & pc : points)
+  {
+    if (!isLastPoint)
+      m_graph.GetOutgoingEdges(Junction(pc, 0 /* altitude */), edges);
+    else
+      m_graph.GetIngoingEdges(Junction(pc, 0 /* altitude */), edges);
+  }
+
+  // Same edges may start on different points if those points are close enough.
+  my::SortUnique(edges, less<Graph::Edge>(), EdgesAreAlmostEqual);
+}
+
+void CandidatePathsGetter::GetAllSuitablePaths(Graph::EdgeVector const & startLines,
+                                               bool const isLastPoint, uint32_t const bearDistM,
+                                               FunctionalRoadClass const frc,
+                                               vector<LinkPtr> & allPaths)
+{
+  ScopedTimer t(m_stats.m_allStatPathsTime);
+
+  queue<LinkPtr> q;
+
+  for (auto const & e : startLines)
+  {
+    auto const u = make_shared<Link>(nullptr /* parent */, e, 0 /* distanceM */);
+    q.push(u);
+  }
+
+  while (!q.empty())
+  {
+    auto const u = q.front();
+    q.pop();
+
+    auto const & currentEdge = u->m_edge;
+    auto const currentEdgeLen = EdgeLength(currentEdge);
+
+    // TODO(mgsergio): Maybe weak this constraint a bit.
+    if (u->m_distanceM + currentEdgeLen >= bearDistM)
+    {
+      allPaths.push_back(u);
+      continue;
+    }
+
+    ASSERT_LESS(u->m_distanceM + currentEdgeLen, bearDistM, ());
+
+    Graph::EdgeVector edges;
+    if (!isLastPoint)
+      m_graph.GetOutgoingEdges(currentEdge.GetEndJunction(), edges);
+    else
+      m_graph.GetIngoingEdges(currentEdge.GetStartJunction(), edges);
+
+    for (auto const & e : edges)
+    {
+      if (EdgesAreAlmostEqual(e.GetReverseEdge(), currentEdge))
+        continue;
+
+      ASSERT(currentEdge.HasRealPart(), ());
+
+      if (!PassesRestriction(e, frc, kFRCThreshold, m_infoGetter))
+        continue;
+
+      // TODO(mgsergio): Should we check form of way as well?
+
+      if (u->IsJunctionInPath(e.GetEndJunction()))
+        continue;
+
+      auto const & p = make_shared<Link>(u, e, u->m_distanceM + currentEdgeLen);
+      q.push(p);
+    }
+  }
+}
+
+void CandidatePathsGetter::GetBestCandidatePaths(
+    vector<LinkPtr> const & allPaths, bool const isLastPoint, uint32_t const requiredBearing,
+    uint32_t const bearDistM, m2::PointD const & startPoint, vector<Graph::EdgeVector> & candidates)
+{
+  ScopedTimer t(m_stats.m_bestCandidatePathsTime);
+
+  set<CandidatePath> candidatePaths;
+  set<CandidatePath> fakeEndingsCandidatePaths;
+
+  BearingPointsSelector pointsSelector(bearDistM, isLastPoint);
+  for (auto const & l : allPaths)
+  {
+    auto const bearStartPoint = pointsSelector.GetBearingStartPoint(l->GetStartEdge());
+    auto const startPointsDsitance =
+        static_cast<uint32_t>(MercatorBounds::DistanceOnEarth(bearStartPoint, startPoint));
+
+    uint32_t traceBackIterationsLeft = 3;
+    for (auto const * part = &l; (*part); part = &((*part)->m_parent))
```

Comment:
optional: for readability, please, change the loop to ```auto part = l; paret; part = part->m_parent```.

Reply:
Done